### PR TITLE
Use whitelist to find go files, run find only once

### PR DESCRIPTION
**Pull request by silverwind**
- Use a whitelist-based approach to find *.go files so we can use standard find syntax. Also included is a change that files no longer pass a leading './' which should help performance further.

- Instead of running `find` multiple times in make because of the lazy evaluation, run it only once and add the bindata files when the bindata tag is present

This is another huge speedup on my machine of around 2000%.

### CODE REVIEW

**Comment by silverwind** [comment]
@guillep2k this should solve what you mentioned yesterday. 

What's interesting is that even if I exclude the bindata files from `GO_SOURCES`, it still seems to be able to build in the bindata. To be sure they are still in the list, I did some extra work to have them included when the tag is present.

File: Makefile [hunk 1]
```diff
@@ -82,13 +72,23 @@ WEBPACK_DEST_DIRS := public/js public/css
 BINDATA_DEST := modules/public/bindata.go modules/options/bindata.go modules/templates/bindata.go
 BINDATA_HASH := $(addsuffix .hash,$(BINDATA_DEST))
 
+TAGS ?=
+TAGS_SPLIT := $(subst $(COMMA), ,$(TAGS))
+TAGS_EVIDENCE := $(MAKE_EVIDENCE_DIR)/tags
+
+GO_SOURCES := $(wildcard *.go)
+GO_SOURCES += $(shell find cmd integrations models modules routers scripts services vendor -type f -name "*.go" -not -path modules/options/bindata.go -not -path modules/public/bindata.go -not -path modules/templates/bindata.go)
```
Comment (guillep2k) [inline Makefile]: I see, if TAGS change we need to get rid of bindata.go. 👍

**Review by guillep2k — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (guillep2k) [inline Makefile]: Maybe it's neater to have:
```
GOSRC_DIRS := cmd integrations models modules routers scripts services vendor
```

Comment (silverwind) [reply Makefile]: Indeed

Comment (silverwind) [reply Makefile]: Yes, it's done this way so it works regardless whether bindata files already exists or not.

**Comment by guillep2k** [comment]
> 
> 
> @guillep2k this should solve what you mentioned yesterday.
> 
> What's interesting is that even if I exclude the bindata files from `GO_SOURCES`, it still seems to be able to build in the bindata. To be sure they are still in the list, I did some extra work to have them included when the tag is present.

It's because the go compiler compiles the whole package. Including them in the source list is only needed for change detection.

**Comment by silverwind** [comment]
One more potential point for optimization is `PACKAGES` which is also lazy-evaluated. Any pointers if this can be made one-time as well?

**Review by 6543 — APPROVED** [review]

File: Makefile [hunk 2]
```diff
@@ -82,13 +72,24 @@ WEBPACK_DEST_DIRS := public/js public/css
...
+	GO_SOURCES += $(BINDATA_DEST)
+endif
+
+GO_SOURCES_OWN := $(filter-out vendor/% %/bindata.go, $(GO_SOURCES))
+
 FOMANTIC_CONFIGS := semantic.json web_src/fomantic/theme.config.less web_src/fomantic/_site/globals/site.variables
 FOMANTIC_DEST := public/fomantic/semantic.min.js public/fomantic/semantic.min.css
 FOMANTIC_DEST_DIR := public/fomantic
 
-TAGS ?=
```
Comment (6543) [inline Makefile:89]: this code move looks unrelated but i consider this as format code ...

**Comment by guillep2k** [comment]
> 
> 
> One more potential point for optimization is `PACKAGES` which is also lazy-evaluated. Any pointers if this can be made one-time as well?

I don't think we need to have it dynamic (but keep on reading). AFAIU, the only reason for keeping it "dynamic" would be to support a forward reference. Consider the following:

```
GO_SOURCES = $(shell find $(SOURCE_DIRS) -name "*.go" -print)

[...]

SOURCE_DIRS = something something something

[...]

goexe: $(GO_SOURCES)
    ....
```

Here `GO_SOURCES` needs to be evaluated later because `SOURCE_DIRS` is not defined at the point `GO_SOURCES` is declared. If you'd use `:=`, then `find` would have no directory list and the results would be invalid.

I don't think that's the situation with `PACKAGES`.

Having said that, `$(PACKAGES)` seems to be used only within specific recipes (i.e. not for rules), like `vet` or `test`, so unless we're executing more than one of those rules in the same run, there will probably be no benefit in changing the definition.

File: Makefile (hunk 2 above)
Comment (silverwind) [reply Makefile:89]: It had to move because `GO_SOURCES` now depends on `TAGS`

**Comment by silverwind** [comment]
Right, `PACKAGES` can be handled in another PR. It does seem like a expensive operation so should bring more speed improvements.

**Comment by silverwind** [comment]
@guillep2k your example contains the anti-pattern of using a variable before it was defined. I would not allow such usage, even with lazy evaluation.

Maybe there are linters for Makefiles that could catch such anti-patterns, a quick search leads me to https://github.com/mrtazz/checkmake.

**Comment by guillep2k** [comment]
> 
> 
> @guillep2k your example contains the anti-pattern of using a variable before it was defined. I would not allow such usage, even with lazy evaluation.
> 

It was just a (lame) example, but quite [a useful pattern in make](https://www.gnu.org/software/make/manual/html_node/Flavors.html#Flavors), because then you've got a find functionality and many places where you can use it in different scenarios. We're using only one Makefile, but [there are build systems](https://stackoverflow.com/a/30215530/225979) with 10s or 100s that chain up and down. As with C++, you _will_ need "foward declarations" at some point. 😄

**Comment by silverwind** [comment]
@jolheiser it seems your review was requested. Would you mind?

**Review by jolheiser — APPROVED** [review]
LGTM

**Comment by guillep2k** [comment]
Is this error coming from the PR itself or is it a glitch?

https://drone.gitea.io/go-gitea/gitea/22479/2/10

image

Although it only failed in MySQL builds, sounds _Makefiley_ to me.

I'm restarting the CI just in case.

**Comment by silverwind** [comment]
`contrib` is not included in `GO_DIRS`. I assume it should?

**Comment by silverwind** [comment]
Added contrib which should get rid of these warnings.

**Comment by silverwind** [comment]
Hmm that didn't do anything to the warnings. They certainly don't come from this PR. I see them on https://drone.gitea.io/go-gitea/gitea/22489/2/10 too.

`contrib` was included in the build before this change, but I deemed it unneccesary. It does contain three go files:

```
contrib/pr/checkout.go
contrib/fixtures/fixture_generation.go
contrib/environment-to-ini/environment-to-ini.go

**Comment by guillep2k** [comment]
I don't think `/contrib` should be part of the main build. There's a `pr\#%:` rule that uses it, though, but that's for manual invocation only.

**Comment by silverwind** [comment]
ok, reverted

**Comment by silverwind** [comment]
rebased, good to go

**Comment by guillep2k** [comment]
Ping LG-TM

**Pull request merged**